### PR TITLE
Allow setting activation_token in supported portals

### PR DESCRIPTION
**Pull request by apricotbucket28**
Allow manually setting the `activation_token` option for the [OpenURI](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.OpenURI.html) and [DynamicLauncher.Launch](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.DynamicLauncher.html#org-freedesktop-portal-dynamiclauncher-launch) portals.

I saw that https://github.com/bilelmoussaoui/ashpd/pull/70 is opened but it also includes additional integration, while this just allows setting the token if we already got it from somewhere else (e.g. directly via the Wayland protocol)

(related commit: https://github.com/apricotbucket28/zed/commit/15489a03966b1c627c9596e0635c5f45c9411b9e)

### CODE REVIEW

File: src/desktop/dynamic_launcher.rs [hunk 1]
```diff
@@ -321,9 +321,15 @@ impl<'a> DynamicLauncherProxy<'a> {
...
-        let options: HashMap<&str, zvariant::Value<'_>> = HashMap::new();
+    pub async fn launch(
+        &self,
+        desktop_file_id: &str,
+        activation_token: impl Into<Option<String>>,
+    ) -> Result<(), Error> {
+        let mut options: HashMap<&str, zvariant::Value<'_>> = HashMap::new();
+        if let Some(activation_token) = activation_token.into() {
+            options.insert("activation_token", activation_token.into());
+        }
```
Comment (bilelmoussaoui) [inline src/desktop/dynamic_launcher.rs:332]: This should be a struct similar to other options we have in the library.

Comment (bilelmoussaoui) [reply src/desktop/dynamic_launcher.rs:332]: But also I am not sure we want to expose the activation token as a random string, I would instead make it a new type that we could potentially validate

Comment (bilelmoussaoui) [reply src/desktop/dynamic_launcher.rs:332]: I know that you can't really validate an activation token under wayland but I don't know if it is the case for x11. Also, we would ideally expose in the future a way to create an activation token from a wayland / x11 surface, similar to how we do for the WindowIdentifier. See #70

**Review by bilelmoussaoui — COMMENTED** [review]
While you are at it, we should also expose the activation_token in the email portal

File: src/desktop/dynamic_launcher.rs (hunk 1 above)
Comment (apricotbucket28) [reply src/desktop/dynamic_launcher.rs:332]: Looking at the validation part, it seems like neither Wayland nor [X11](https://github.com/rust-windowing/winit/blob/db2c97a995bc403e0442675c7364010b806d5431/src/platform_impl/linux/x11/activation.rs#L19) have a unified standard for the token. Should I still add an `ActivationToken` type anyway?

Comment (bilelmoussaoui) [reply src/desktop/dynamic_launcher.rs]: I would still define a new type as we would need to provide constructors and so on for later. For the validation part, it would be nice to document that both x11 and wayland activation tokens can't be validated, so just implement From<String> and other traits.

File: src/desktop/email.rs [hunk 2]
```diff
@@ -177,11 +176,9 @@ impl EmailRequest {
         self
     }
 
-    // TODO Added in version 4 of the interface.
```
Comment (bilelmoussaoui) [inline src/desktop/email.rs:180]: Please keep this todo bit, as we would like to potentially only send stuff that are supported by the portal (micro-optimization, probably worthless but oh well)

File: src/desktop/open_uri.rs [hunk 3]
```diff
@@ -171,6 +171,13 @@ impl OpenFileRequest {
         self
     }
 
+    /// Sets the token that can be used to activate the chosen application.
```
Comment (bilelmoussaoui) [inline src/desktop/open_uri.rs:174]: I see that you are only using this in your zed PR, please split it in a separate commit so that I could backport it for you otherwise you would have to wait for a major release in few weeks

Comment (apricotbucket28) [reply src/desktop/open_uri.rs:174]: Appreciate the concerns, but I think that Zed could switch to the git repo for now (Zed on Linux is still in development) :slightly_smiling_face:

File: src/activation_token/mod.rs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+///
+/// No guarantees are made for the token structure.
+#[derive(Debug, Deserialize, Serialize, Type, PartialEq, Eq, Hash, Clone)]
+pub struct ActivationToken(String);
+
+impl From<String> for ActivationToken {
+    fn from(value: String) -> Self {
+        ActivationToken(value)
+    }
+}
```
Comment (bilelmoussaoui) [inline src/activation_token/mod.rs:16]: would you mind adding a `From<&str>` variant as well? for ease of use

File: src/activation_token/mod.rs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+/// A token that can be used to activate an application.
+///
+/// No guarantees are made for the token structure.
+#[derive(Debug, Deserialize, Serialize, Type, PartialEq, Eq, Hash, Clone)]
+pub struct ActivationToken(String);
+
+impl From<String> for ActivationToken {
+    fn from(value: String) -> Self {
+        ActivationToken(value)
```
Comment (bilelmoussaoui) [inline src/activation_token/mod.rs:14]:
```suggestion
        Self(value)
```

**Review by bilelmoussaoui — APPROVED** [review]
otherwise lgtm

**Comment by bilelmoussaoui** [comment]
Thanks a lot!
